pipettes/firmware/i2c.c

```diff
@@ -53,3 +112,65 @@ I2C_HandleTypeDef* MX_I2C_Init()
     return &hi2c;
 
 }
+
+I2C_HandleTypeDef* MX_I2C3_Init()
+{
+    hi2c3.Instance = I2C3;
+    hi2c3.Init.Timing = 0x30A0A7FB;
+    hi2c3.Init.OwnAddress1 = 0;
+    hi2c3.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+    hi2c3.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+    hi2c3.Init.OwnAddress2 = 0;
+    hi2c3.Init.OwnAddress2Masks = I2C_OA2_NOMASK;
+    hi2c3.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
+    hi2c3.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+    if (HAL_I2C_Init(&hi2c3) != HAL_OK)
+    {
+        Error_Handler();
+    }
+    /** Configure Analogue filter
+    */
+    if (HAL_I2CEx_ConfigAnalogFilter(&hi2c3, I2C_ANALOGFILTER_ENABLE) != HAL_OK)
+    {
+        Error_Handler();
+    }
+    /** Configure Digital filter
+    */
+    if (HAL_I2CEx_ConfigDigitalFilter(&hi2c3, 0) != HAL_OK)
+    {
+        Error_Handler();
+    }
+
+    return &hi2c3;
+
+}
+
+int data_ready() {
+    return HAL_GPIO_ReadPin(GPIOC, GPIO_PIN_9) == GPIO_PIN_SET;
+}
+
+int tip_present() {
+    return HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_10) == GPIO_PIN_SET;
+}
+
+void enable_eeprom() {
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+}
+
+void disable_eeprom() {
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+}
+
+I2CHandlerStruct i2c_setup(PipetteType pipette_type) {
```

Comment:
It is very un-C like to pass or return structs by value. 

I realize this function is only called once so this isn't a big deal, but for the purpose of writing in the C idiom I suggest you rewrite this function to accept a pointer to a `I2CHandlerStruct`.

Reply:
That makes sense! Generally, is there anything you would want to return by value aside from a primitive type?

Reply:
In C, just primitives.

Reply:
You're right but also that's a thing I've always hated about C, and it's kind of held over from older implementations and standards. I'm ok with either way.